// @wordpress/block-library/build/query/edit/query-inspector-controls.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = QueryInspectorControls;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

var _data = require("@wordpress/data");

var _utils = require("../utils");

var _constants = require("../constants");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function QueryInspectorControls(_ref) {
  var _categories$terms, _tags$terms;

  var query = _ref.query,
      setQuery = _ref.setQuery;
  var order = query.order,
      orderBy = query.orderBy,
      selectedAuthorId = query.author,
      postType = query.postType;

  var _useState = (0, _element.useState)(true),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showCategories = _useState2[0],
      setShowCategories = _useState2[1];

  var _useState3 = (0, _element.useState)(true),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      showTags = _useState4[0],
      setShowTags = _useState4[1];

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _getPostTypes;

    var _select = select('core'),
        getEntityRecords = _select.getEntityRecords,
        getPostTypes = _select.getPostTypes;

    var termsQuery = {
      per_page: _constants.MAX_FETCHED_TERMS
    };

    var _categories = getEntityRecords('taxonomy', 'category', termsQuery);

    var _tags = getEntityRecords('taxonomy', 'post_tag', termsQuery);

    var excludedPostTypes = ['attachment'];
    var filteredPostTypes = (_getPostTypes = getPostTypes()) === null || _getPostTypes === void 0 ? void 0 : _getPostTypes.filter(function (_ref2) {
      var viewable = _ref2.viewable,
          slug = _ref2.slug;
      return viewable && !excludedPostTypes.includes(slug);
    });
    return {
      categories: (0, _utils.getTermsInfo)(_categories),
      tags: (0, _utils.getTermsInfo)(_tags),
      authorList: getEntityRecords('root', 'user', {
        per_page: -1
      }),
      postTypes: filteredPostTypes
    };
  }, []),
      authorList = _useSelect.authorList,
      categories = _useSelect.categories,
      tags = _useSelect.tags,
      postTypes = _useSelect.postTypes;

  var postTypesTaxonomiesMap = (0, _element.useMemo)(function () {
    if (!(postTypes === null || postTypes === void 0 ? void 0 : postTypes.length)) return;
    return postTypes.reduce(function (accumulator, type) {
      accumulator[type.slug] = type.taxonomies;
      return accumulator;
    }, {});
  }, [postTypes]);
  (0, _element.useEffect)(function () {
    if (!postTypesTaxonomiesMap) return;
    var postTypeTaxonomies = postTypesTaxonomiesMap[postType];
    setShowCategories(postTypeTaxonomies.includes('category'));
    setShowTags(postTypeTaxonomies.includes('post_tag'));
  }, [postType, postTypesTaxonomiesMap]);
  var postTypesSelectOptions = (0, _element.useMemo)(function () {
    return (postTypes || []).map(function (_ref3) {
      var labels = _ref3.labels,
          slug = _ref3.slug;
      return {
        label: labels.singular_name,
        value: slug
      };
    });
  }, [postTypes]);

  var onPostTypeChange = function onPostTypeChange(newValue) {
    var updateQuery = {
      postType: newValue
    };

    if (!postTypesTaxonomiesMap[newValue].includes('category')) {
      updateQuery.categoryIds = [];
    }

    if (!postTypesTaxonomiesMap[newValue].includes('post_tag')) {
      updateQuery.tagIds = [];
    }

    setQuery(updateQuery);
  }; // Handles categories and tags changes.


  var onTermsChange = function onTermsChange(terms, queryProperty) {
    return function (newTermValues) {
      var termIds = newTermValues.reduce(function (accumulator, termValue) {
        var _terms$mapByName$term;

        var termId = (termValue === null || termValue === void 0 ? void 0 : termValue.id) || ((_terms$mapByName$term = terms.mapByName[termValue]) === null || _terms$mapByName$term === void 0 ? void 0 : _terms$mapByName$term.id);
        if (termId) accumulator.push(termId);
        return accumulator;
      }, []);
      setQuery((0, _defineProperty2.default)({}, queryProperty, termIds));
    };
  };

  var onCategoriesChange = onTermsChange(categories, 'categoryIds');
  var onTagsChange = onTermsChange(tags, 'tagIds');

  var _useState5 = (0, _element.useState)(query.search),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      querySearch = _useState6[0],
      setQuerySearch = _useState6[1];

  var onChangeDebounced = (0, _element.useCallback)((0, _lodash.debounce)(function () {
    return setQuery({
      search: querySearch
    });
  }, 250), [querySearch]);
  (0, _element.useEffect)(function () {
    onChangeDebounced();
    return onChangeDebounced.cancel;
  }, [querySearch, onChangeDebounced]);
  return (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Filtering and Sorting')
  }, (0, _element.createElement)(_components.SelectControl, {
    options: postTypesSelectOptions,
    value: postType,
    label: (0, _i18n.__)('Post Type'),
    onChange: onPostTypeChange
  }), showCategories && (categories === null || categories === void 0 ? void 0 : (_categories$terms = categories.terms) === null || _categories$terms === void 0 ? void 0 : _categories$terms.length) > 0 && (0, _element.createElement)(_components.FormTokenField, {
    label: (0, _i18n.__)('Categories'),
    value: (query.categoryIds || []).map(function (categoryId) {
      return {
        id: categoryId,
        value: categories.mapById[categoryId].name
      };
    }),
    suggestions: categories.names,
    onChange: onCategoriesChange
  }), showTags && (tags === null || tags === void 0 ? void 0 : (_tags$terms = tags.terms) === null || _tags$terms === void 0 ? void 0 : _tags$terms.length) > 0 && (0, _element.createElement)(_components.FormTokenField, {
    label: (0, _i18n.__)('Tags'),
    value: (query.tagIds || []).map(function (tagId) {
      return {
        id: tagId,
        value: tags.mapById[tagId].name
      };
    }),
    suggestions: tags.names,
    onChange: onTagsChange
  }), (0, _element.createElement)(_components.QueryControls, (0, _extends2.default)({
    order: order,
    orderBy: orderBy,
    selectedAuthorId: selectedAuthorId,
    authorList: authorList
  }, {
    onOrderChange: function onOrderChange(value) {
      return setQuery({
        order: value
      });
    },
    onOrderByChange: function onOrderByChange(value) {
      return setQuery({
        orderBy: value
      });
    },
    onAuthorChange: function onAuthorChange(value) {
      return setQuery({
        author: value !== '' ? +value : undefined
      });
    }
  })), (0, _element.createElement)(_components.TextControl, {
    label: (0, _i18n.__)('Search'),
    value: querySearch,
    onChange: setQuerySearch
  })));
}
