// @wordpress/interface/build-module/components/action-item/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { isEmpty, noop } from 'lodash';
/**
 * WordPress dependencies
 */

import { ButtonGroup, Button, Slot, Fill } from '@wordpress/components';

function ActionItemSlot(_ref) {
  var name = _ref.name,
      _ref$as = _ref.as,
      as = _ref$as === void 0 ? [ButtonGroup, Button] : _ref$as,
      _ref$fillProps = _ref.fillProps,
      fillProps = _ref$fillProps === void 0 ? {} : _ref$fillProps,
      bubblesVirtually = _ref.bubblesVirtually,
      props = _objectWithoutProperties(_ref, ["name", "as", "fillProps", "bubblesVirtually"]);

  var _as = _slicedToArray(as, 2),
      Container = _as[0],
      Item = _as[1];

  return createElement(Slot, {
    name: name,
    bubblesVirtually: bubblesVirtually,
    fillProps: _objectSpread({
      as: Item
    }, fillProps)
  }, function (fills) {
    return !isEmpty(fills) && createElement(Container, props, fills);
  });
}

function ActionItem(_ref2) {
  var name = _ref2.name,
      as = _ref2.as,
      onClick = _ref2.onClick,
      props = _objectWithoutProperties(_ref2, ["name", "as", "onClick"]);

  return createElement(Fill, {
    name: name
  }, function (fillProps) {
    var fpOnClick = fillProps.onClick,
        fpAs = fillProps.as;
    var Item = as || fpAs || Button;
    return createElement(Item, _extends({
      onClick: onClick || fpOnClick ? function () {
        (onClick || noop).apply(void 0, arguments);
        (fpOnClick || noop).apply(void 0, arguments);
      } : undefined
    }, props));
  });
}

ActionItem.Slot = ActionItemSlot;
export default ActionItem;
