// @wordpress/edit-post/build-module/components/header/fullscreen-mode-close/index.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { get } from 'lodash';
/**
 * WordPress dependencies
 */

import { useSelect } from '@wordpress/data';
import { Button, Icon } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { addQueryArgs } from '@wordpress/url';
import { wordpress } from '@wordpress/icons';

function FullscreenModeClose(_ref) {
  var showTooltip = _ref.showTooltip;

  var _useSelect = useSelect(function (select) {
    var _select = select('core/editor'),
        getCurrentPostType = _select.getCurrentPostType;

    var _select2 = select('core/edit-post'),
        isFeatureActive = _select2.isFeatureActive;

    var _select3 = select('core/data'),
        isResolving = _select3.isResolving;

    var _select4 = select('core'),
        getEntityRecord = _select4.getEntityRecord,
        getPostType = _select4.getPostType;

    var siteData = getEntityRecord('root', '__unstableBase', undefined) || {};
    return {
      isActive: isFeatureActive('fullscreenMode'),
      isRequestingSiteIcon: isResolving('core', 'getEntityRecord', ['root', '__unstableBase', undefined]),
      postType: getPostType(getCurrentPostType()),
      siteIconUrl: siteData.site_icon_url
    };
  }, []),
      isActive = _useSelect.isActive,
      isRequestingSiteIcon = _useSelect.isRequestingSiteIcon,
      postType = _useSelect.postType,
      siteIconUrl = _useSelect.siteIconUrl;

  if (!isActive || !postType) {
    return null;
  }

  var buttonIcon = createElement(Icon, {
    size: "36px",
    icon: wordpress
  });

  if (siteIconUrl) {
    buttonIcon = createElement("img", {
      alt: __('Site Icon'),
      className: "edit-post-fullscreen-mode-close_site-icon",
      src: siteIconUrl
    });
  } else if (isRequestingSiteIcon) {
    buttonIcon = null;
  }

  return createElement(Button, {
    className: "edit-post-fullscreen-mode-close has-icon",
    href: addQueryArgs('edit.php', {
      post_type: postType.slug
    }),
    label: get(postType, ['labels', 'view_items'], __('Back')),
    showTooltip: showTooltip
  }, buttonIcon);
}

export default FullscreenModeClose;
