// @wordpress/edit-post/build/components/layout/popover-wrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PopoverWrapper;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _components = require("@wordpress/components");

var _keycodes = require("@wordpress/keycodes");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function stopPropagation(event) {
  event.stopPropagation();
}

var DetectOutside = (0, _components.withFocusOutside)( /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(_class, _Component);

  var _super = _createSuper(_class);

  function _class() {
    (0, _classCallCheck2.default)(this, _class);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(_class, [{
    key: "handleFocusOutside",
    value: function handleFocusOutside(event) {
      this.props.onFocusOutside(event);
    }
  }, {
    key: "render",
    value: function render() {
      return this.props.children;
    }
  }]);
  return _class;
}(_element.Component));
var FocusManaged = (0, _components.withConstrainedTabbing)((0, _components.withFocusReturn)(function (_ref) {
  var children = _ref.children;
  return children;
}));

function PopoverWrapper(_ref2) {
  var onClose = _ref2.onClose,
      children = _ref2.children,
      className = _ref2.className;

  // Event handlers
  var maybeClose = function maybeClose(event) {
    // Close on escape
    if (event.keyCode === _keycodes.ESCAPE && onClose) {
      event.stopPropagation();
      onClose();
    }
  }; // Disable reason: this stops certain events from propagating outside of the component.
  //   - onMouseDown is disabled as this can cause interactions with other DOM elements

  /* eslint-disable jsx-a11y/no-static-element-interactions */


  return (0, _element.createElement)("div", {
    className: className,
    onKeyDown: maybeClose,
    onMouseDown: stopPropagation
  }, (0, _element.createElement)(DetectOutside, {
    onFocusOutside: onClose
  }, (0, _element.createElement)(FocusManaged, null, children)));
  /* eslint-enable jsx-a11y/no-static-element-interactions */
}
