// @wordpress/edit-post/build/components/preferences-modal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreferencesModal = PreferencesModal;
exports.default = void 0;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _editor = require("@wordpress/editor");

var _section = _interopRequireDefault(require("./section"));

var _options = require("./options");

var _metaBoxesSection = _interopRequireDefault(require("./meta-boxes-section"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var MODAL_NAME = 'edit-post/preferences';

function PreferencesModal(_ref) {
  var isModalActive = _ref.isModalActive,
      isViewable = _ref.isViewable,
      closeModal = _ref.closeModal;

  if (!isModalActive) {
    return null;
  }

  return (0, _element.createElement)(_components.Modal, {
    className: "edit-post-preferences-modal",
    title: (0, _i18n.__)('Preferences'),
    closeLabel: (0, _i18n.__)('Close'),
    onRequestClose: closeModal
  }, (0, _element.createElement)(_section.default, {
    title: (0, _i18n.__)('General')
  }, (0, _element.createElement)(_options.EnablePublishSidebarOption, {
    help: (0, _i18n.__)('Review settings such as categories and tags.'),
    label: (0, _i18n.__)('Include pre-publish checklist')
  }), (0, _element.createElement)(_options.EnableFeature, {
    featureName: "mostUsedBlocks",
    help: (0, _i18n.__)('Places the most frequent blocks in the block library.'),
    label: (0, _i18n.__)('Show most used blocks')
  })), (0, _element.createElement)(_section.default, {
    title: (0, _i18n.__)('Keyboard')
  }, (0, _element.createElement)(_options.EnableFeature, {
    featureName: "keepCaretInsideBlock",
    help: (0, _i18n.__)('Aids screen readers by stopping text caret from leaving blocks.'),
    label: (0, _i18n.__)('Contain text cursor inside block')
  })), (0, _element.createElement)(_section.default, {
    title: (0, _i18n.__)('Appearance')
  }, (0, _element.createElement)(_options.EnableFeature, {
    featureName: "reducedUI",
    help: (0, _i18n.__)('Compacts options and outlines in the toolbar.'),
    label: (0, _i18n.__)('Reduce the interface')
  }), (0, _element.createElement)(_options.EnableFeature, {
    featureName: "focusMode",
    help: (0, _i18n.__)('Highlights the current block and fades other content.'),
    label: (0, _i18n.__)('Spotlight mode')
  }), (0, _element.createElement)(_options.EnableFeature, {
    featureName: "showIconLabels",
    help: (0, _i18n.__)('Shows text instead of icons in toolbar.'),
    label: (0, _i18n.__)('Display button labels')
  }), (0, _element.createElement)(_options.EnableFeature, {
    featureName: "themeStyles",
    help: (0, _i18n.__)('Make the editor look like your theme.'),
    label: (0, _i18n.__)('Use theme styles')
  })), (0, _element.createElement)(_section.default, {
    title: (0, _i18n.__)('Document settings'),
    description: (0, _i18n.__)('Choose what displays in the panel.')
  }, (0, _element.createElement)(_options.EnablePluginDocumentSettingPanelOption.Slot, null), isViewable && (0, _element.createElement)(_options.EnablePanelOption, {
    label: (0, _i18n.__)('Permalink'),
    panelName: "post-link"
  }), (0, _element.createElement)(_editor.PostTaxonomies, {
    taxonomyWrapper: function taxonomyWrapper(content, taxonomy) {
      return (0, _element.createElement)(_options.EnablePanelOption, {
        label: (0, _lodash.get)(taxonomy, ['labels', 'menu_name']),
        panelName: "taxonomy-panel-".concat(taxonomy.slug)
      });
    }
  }), (0, _element.createElement)(_editor.PostFeaturedImageCheck, null, (0, _element.createElement)(_options.EnablePanelOption, {
    label: (0, _i18n.__)('Featured image'),
    panelName: "featured-image"
  })), (0, _element.createElement)(_editor.PostExcerptCheck, null, (0, _element.createElement)(_options.EnablePanelOption, {
    label: (0, _i18n.__)('Excerpt'),
    panelName: "post-excerpt"
  })), (0, _element.createElement)(_editor.PostTypeSupportCheck, {
    supportKeys: ['comments', 'trackbacks']
  }, (0, _element.createElement)(_options.EnablePanelOption, {
    label: (0, _i18n.__)('Discussion'),
    panelName: "discussion-panel"
  })), (0, _element.createElement)(_editor.PageAttributesCheck, null, (0, _element.createElement)(_options.EnablePanelOption, {
    label: (0, _i18n.__)('Page attributes'),
    panelName: "page-attributes"
  }))), (0, _element.createElement)(_metaBoxesSection.default, {
    title: (0, _i18n.__)('Additional panels'),
    description: (0, _i18n.__)('Add extra areas to the editor.')
  }));
}

var _default = (0, _compose.compose)((0, _data.withSelect)(function (select) {
  var _select = select('core/editor'),
      getEditedPostAttribute = _select.getEditedPostAttribute;

  var _select2 = select('core'),
      getPostType = _select2.getPostType;

  var postType = getPostType(getEditedPostAttribute('type'));
  return {
    isModalActive: select('core/edit-post').isModalActive(MODAL_NAME),
    isViewable: (0, _lodash.get)(postType, ['viewable'], false)
  };
}), (0, _data.withDispatch)(function (dispatch) {
  return {
    closeModal: function closeModal() {
      return dispatch('core/edit-post').closeModal();
    }
  };
}))(PreferencesModal);

exports.default = _default;
